pypeit/reduce.py

```diff
@@ -190,64 +143,213 @@ def parse_manual_dict(self, manual_dict, neg=False):
+
+        Returns:
+            SpecObj:
+
+        """
+        pass
+
+    def run(self, std_trace=None, manual_extract_dict=None, show_peaks=False):
+        """
+        Primary code flow for PypeIt reductions
+
+        Args:
+            std_trace (ndarray, optional):
+                Trace of the standard star
+            manual_extract_dict (dict, optional):
+            show_peaks (bool, optional):
+                Show peaks in find_objects methods
+
+        Returns:
+            tuple: skymodel (ndarray), objmodel (ndarray), ivarmodel (ndarray),
+               outmask (ndarray), sobjs (SpecObjs).  See main doc string for description
+
+        """
+        # First pass object finding
+        self.sobjs_obj, self.nobj, skymask_init = \
+            self.find_objects(self.sciImg.image, std_trace=std_trace,
+                              show_peaks=show_peaks,
+                              show=self.reduce_show & (not self.std_redux),
+                              manual_extract_dict=manual_extract_dict)
+
+        # Global sky subtract
+        self.initial_sky = \
+            self.global_skysub(skymask=skymask_init).copy()
+
+        # Second pass object finding
+        if (not self.std_redux) and (not self.par['scienceimage']['findobj']['skip_second_find']):
+            self.sobjs_obj, self.nobj, self.skymask = \
+                self.find_objects(self.sciImg.image - self.initial_sky,
+                                  std_trace=std_trace, show=self.reduce_show,
+                                  show_peaks=show_peaks,
+                                  manual_extract_dict=manual_extract_dict)
+        else:
+            msgs.info("Skipping 2nd run of finding objects")
+
+        # Do we have any objects to proceed with?
+        if self.sobjs_obj.nobj > 0:
+            # Global sky subtraction second pass. Uses skymask from object finding
+            if (self.std_redux or self.par['scienceimage']['extraction']['skip_optimal'] or
+                    self.par['scienceimage']['findobj']['skip_second_find']):
+                self.global_sky = self.initial_sky.copy()
+            else:
+                self.global_sky = self.global_skysub(skymask=self.skymask,
+                                                     show=self.reduce_show)
+            # Extract + Return
+            self.skymodel, self.objmodel, self.ivarmodel, self.outmask, self.sobjs = self.extract(self.global_sky,
+                                                                                                  self.sobjs_obj)
+        else:  # No objects, pass back what we have
+            self.skymodel = self.initial_sky
+            self.objmodel = np.zeros_like(self.sciImg.image)
+            # Set to sciivar. Could create a model but what is the point?
+            self.ivarmodel = np.copy(self.sciImg.ivar)
+            # Set to the initial mask in case no objects were found
+            self.outmask = self.sciImg.mask
```

Comment:
Something is inconsistent here and you need to consult the older code. In an IR reduction there could be only negative objects but no positive objects. I'm now confused about the order of purging and assignment below and whether you have mixed it up. The other point is whether self.sobjs_obj.nobj = 0 for positive objects, do we still want to reduce and extract? I think the answer is no, but not sure what we were doing before.

Reply:
Indeed this should act on `self.sobjs`, not `self.sobjs_obj`.  
Carrying these two around has confused me..

As for the original code, we did have a purge inside the
block with `nobj=0` and only working on `self.sobjs_obj`.
But as there are no objects, that wasn't doing anything.

Reply:
deal with.